File: lesson-3_dynamic-programming/0-1-knapsack-problem/memoization/main.cpp
#include <iostream>
#include <vector>
#include <fmt/core.h>

int getFav(int i,
           int balance,
           const std::vector<int> &values,
           const std::vector<int> &like,
           std::vector<std::vector<int>> &rec
);

int main() {
    int n, balance;
    std::cin >> n >> balance;

    auto values = std::vector<int>(n);
    auto like = std::vector<int>(n);
    auto rec = std::vector<std::vector<int>>(n);
    for (int i = 0; i < n; i++) {
        std::cin >> values[i] >> like[i];
        rec[i] = std::vector<int>(balance + 1);
    }
    fmt::print("{}\n", getFav(n - 1, balance, values, like, rec));
    return 0;
}

int getFav(
        int i,
        int balance,
        const std::vector<int> &values,
        const std::vector<int> &like,
        std::vector<std::vector<int>> &rec
) {
    if (i == -1 || balance <= 0) {
        return 0;
    }
    if (rec[i][balance] == 0) {
        if (values[i] > balance) {
            rec[i][balance] = getFav(i - 1, balance, values, like, rec);
        } else {
            rec[i][balance] = std::max(
                    getFav(i - 1, balance, values, like, rec),
                    getFav(i - 1, balance - values[i], values, like, rec) + like[i]
            );
        }
    }
    return rec[i][balance];
}

File: lesson-3_dynamic-programming/0-1-knapsack-problem/memoization/main.go
package main

import (
    "fmt"
    "math"
)

func main() {
    var n, balance int
    fmt.Scan(&n, &balance)

    values := make([]int, n)
    like := make([]int, n)
    rec := make([][]int, n)
    for i := 0; i < n; i++ {
        fmt.Scan(&values[i], &like[i])
        rec[i] = make([]int, balance+1)
    }
    fmt.Println(getFav(n-1, balance, values, like, rec))
}

func getFav(i, balance int, values, like []int, rec [][]int) int {
    if i == -1 || balance <= 0 {
        return 0
    }
    if rec[i][balance] == 0 {
        if values[i] > balance {
            rec[i][balance] = getFav(i-1, balance, values, like, rec)
        } else {
            rec[i][balance] = int(math.Max(
                float64(getFav(i-1, balance, values, like, rec)),
                float64(getFav(i-1, balance-values[i], values, like, rec)+like[i]),
            ))
        }
    }
    return rec[i][balance]
}
